### PR TITLE
Add spec up spec

### PR DESCRIPTION
**Pull request by decentralgabe**
**work in progress

### CODE REVIEW

**Comment by decentralgabe** [comment]
It might be easiest for you to review this if you check out the branch and render the spec yourself.

You can do that with ...

```sh
git fetch
git checkout origin/spec-up
cd spec
npm i
npm run build
open build/index.html
```

File: spec/spec.md [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+## Introduction
+
+### Background
+
+The evolution of the World Wide Web has been marked by significant shifts in how users interact with online content and services:
+
+1. **Web1**: Read-only, static websites
+2. **Web2**: Read-write, interactive platforms
+3. **Web3**: Read-write-own, blockchain-based applications
+4. **Web5**: Decentralized, user-centric internet, without the need for a blockchain
```
Comment (nitro-neal) [inline spec/spec.md:43]: nice

File: spec/spec.md [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+In-memory signing using [[ref:secp256k1]] ****MUST**** produce k-deterministic low-s signatures with [ECDSA](https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm) as per [[spec:RFC6979]]. Verification ****MUST NOT**** require low-s signatures.
+:::
+
+### Key Management
+
+Web5 implementations ****MUST**** provide a consistent and extensible _public_ interface for key management, with the following minimum concrete implementations:
+
+1. Key Manager Interface
+2. In-Memory Key Manager
+3. AWS KMS
```
Comment (nitro-neal) [inline spec/spec.md:116]: are we 100% sure we need this / its required for the spec?

File: spec/spec.md [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+::: todo
+Provide detailed specifications for each key management implementation, including APIs and usage guidelines.
+:::
+
+Further, the key manager interface ****MUST**** be passed as an argument to all public API methods that require key material such as:
+* DID Creation
+* Data Signing
+
+::: note
+[AWS KMS](https://docs.aws.amazon.com/kms/latest/developerguide/asymmetric-key-specs.html#key-spec-ecc) does not support [[ref:Ed25519]]. At some point in the future, we can consider introducing support for key wrapping.
```
Comment (nitro-neal) [inline spec/spec.md:129]: especially since it wont event work lol

File: spec/spec.md [hunk 4]
```diff
@@ -0,0 +1,601 @@
...
+:::
+
+Consumers of conformant [[ref:Web5 SDKs]] ****SHOULD**** be able to provide their own `KeyManager` impelementations, if desired.
+
+### JSON Schema
+
+[[ref:JSON-SCHEMA]] support is ****REQUIRED**** for multiple features throughout [[ref:Web5 SDKs]], including, but not limited to support with [[ref:verifiable credentials]], [VC JSON Schemas](#vc-json-schema), and for use in validating the data models this specification has defined.
+
+Conformant implementations of [[ref:Web5 SDKs]] ****MUST**** support at least [[ref:JSON-SCHEMA-DRAFT-7]] and [[ref:JSON-SCHEMA-2020-12]].
```
Comment (nitro-neal) [inline spec/spec.md:138]: Chad says JSON-SCHEMA-2020-12 is backwards compatible with JSON-SCHEMA-DRAFT-7, so just writing in JSON-SCHEMA-DRAFT-7 should be sufficient right?

File: spec/spec.md [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+4. **Resolvability**: [[ref:DIDs]] can be resolved to retrieve associated [[ref:DID Documents]] containing cryptographic material, service endpoints, and other data.
+4. **Cryptographic Verifiability**: [[ref:DIDs]] enable cryptographic verification of claims and interactions.
+
+#### Supported DID Methods
+
+Web5 supports the following DID methods:
+
+| Method | Creation | Resolution | Note |
+|--------|----------|------------|------|
+| [[ref:did:web]] | ❌ | ✅ | - |
```
Comment (nitro-neal) [inline spec/spec.md:162]: We can create a did web document (have this in the cli).. but yea they would have to host it on their domain, so the philosophical question: is a did:web created if its not hosted on a domain?

File: spec/spec.md [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+2. **Control**: The controller has full say over their [[ref:DID]] and associated [[ref:DID Document]].
+3. **Persistence**: [[ref:DIDs]] are persistent identifiers that do not require the continued operation of an underlying organization.
+4. **Resolvability**: [[ref:DIDs]] can be resolved to retrieve associated [[ref:DID Documents]] containing cryptographic material, service endpoints, and other data.
+4. **Cryptographic Verifiability**: [[ref:DIDs]] enable cryptographic verification of claims and interactions.
+
+#### Supported DID Methods
+
+Web5 supports the following DID methods:
+
+| Method | Creation | Resolution | Note |
```
Comment (nitro-neal) [inline spec/spec.md:160]: add update column here?

File: spec/spec.md (hunk 4 above)
Comment (nitro-neal) [reply spec/spec.md:138]: And if someone used features from JSON-SCHEMA-2020-12 then it may or may not be compatible with SCHEMA-DRAFT-7... lol so it seems like everyone will / should only use SCHEMA-DRAFT-7

File: spec/spec.md [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+#### VC Formats
+
+Web5 supports the following credential formats:
+
+1. **JWT**: Verifiable Credentials and Presentations secured as JSON Web Tokens (JWTs) [[spec:RFC7519]] using the `vc-jwt` and `vp-jwt` types.
+2. **JSON-LD**: Support for [[ref:JSON-LD]] context in VCs, although [[ref:JSON-LD]] processing is not performed.
+
+#### Credential Status
+
+Conformant [[ref:Web5 SDKs]] ****MUST**** support Credential Status as specified by [[ref:STATUS-LIST-2021]] for usage with the W3C Verifiable Credential Data Model 1.1 [[ref:VC-DATA-MODEL-1.1]].
```
Comment (nitro-neal) [inline spec/spec.md:238]: not using the updated one? https://www.w3.org/TR/vc-bitstring-status-list/

File: spec/spec.md (hunk 2 above)
Comment (decentralgabe) [reply spec/spec.md:116]: I copied from the current readme I'd love to kill this or just define a key manager interface

File: spec/spec.md (hunk 3 above)
Comment (decentralgabe) [reply spec/spec.md:129]: hmm can you expand?

File: spec/spec.md (hunk 4 above)
Comment (decentralgabe) [reply spec/spec.md:138]: we should be using 2020-12 since it's the latest draft, which is being used as the input document to a LTS type draft. 

more info [here](https://json-schema.org/blog/posts/future-of-json-schema) and [here](https://json-schema.org/blog/posts/the-last-breaking-change) which notes that if you're using 2020-12 you'll be fine (we're not using breaking features)

it should be straightforward to convert our schemas to 2020-12, I've done so in the past using this tool https://github.com/sourcemeta/alterschema

File: spec/spec.md (hunk 6 above)
Comment (decentralgabe) [reply spec/spec.md:160]: good catch

File: spec/spec.md (hunk 7 above)
Comment (decentralgabe) [reply spec/spec.md:238]: there are some weird things that will break if we do, mostly incompatibility with the 1.1 data model since bitstring is designed to be used with 2.0

File: spec/spec.md [hunk 8]
```diff
@@ -0,0 +1,601 @@
...
+~ [Tom Daffrun](https://github.com/tomdaffurn)
+
+**Participate**:
+~ [GitHub repo](https://github.com/TBD54566975/web5-spec)
+~ [File a bug](https://github.com/TBD54566975/web5-spec/issues)
+~ [Commit history](https://github.com/TBD54566975/web5-spec/commits/main)
+
+## Abstract
+
+Web5 is a decentralized web platform that empowers users with control over their data and identity. Built on open standards and protocols, Web5 enables the integration of decentralized identity into applications, allowing developers to create user-centric experiences while ensuring individual ownership and control over personal information. This specification selects standards from the digital identity and implemenets sensible constraints to enable functional, practical, and interoperable implementations of these standards. Following this ethos, the Web5 Specification defines the core components, protocols, and APIs that constitute the Web5 ecosystem, enabling interoperable and privacy-preserving decentralized applications.
```
Comment (KendallWeihe) [inline spec/spec.md:32]:
```suggestion
Web5 is a decentralized web platform that empowers users with control over their data and identity. Built on open standards and protocols, Web5 enables the integration of decentralized identity into applications, allowing developers to create user-centric experiences while ensuring individual ownership and control over personal information. This specification selects standards from the digital identity and implements sensible constraints to enable functional, practical, and interoperable implementations of these standards. Following this ethos, the Web5 Specification defines the core components, protocols, and APIs that constitute the Web5 ecosystem, enabling interoperable and privacy-preserving decentralized applications.
```

File: spec/spec.md [hunk 9]
```diff
@@ -0,0 +1,601 @@
...
+The Web5 SDK provides a comprehensive set of features for building decentralized protocols and applications. The following section outline the core features required to build a conformant [[ref:Web5 SDK]].
+
+### Cryptographic Primitives
+
+Web5 supports the following cryptographic key types for use with the noted corresponding digital signature algorithms:
+
+| Key Type | Algorithm | Function |
+|----------|-----------|----------|
+| [[ref:secp256k1]] | `ES256K` [[spec:RFC8812]] | Signing and Verification |
+| [[ref:Ed25519]] | `EdDSA` [[spec:RFC8032]] | Signing and Verification |
```
Comment (KendallWeihe) [inline spec/spec.md:104]: FWIW I was planning on adding support for the four key types from here https://did-dht.com/registry/index.html#key-type-index

File: spec/spec.md (hunk 2 above)
Comment (KendallWeihe) [reply spec/spec.md:116]: I feel strongly we should kill this as well as (4) and (5) below. I do feel that we need to add support for the 3 I'm proposing removing here, but I lack confidence the 3 belong here in this place at this time. Perhaps we could add a `:::note` with commentary which describes our intention of covering all major platforms' key management at a later date.

File: spec/spec.md [hunk 10]
```diff
@@ -0,0 +1,601 @@
...
+
+::: note
+[AWS KMS](https://docs.aws.amazon.com/kms/latest/developerguide/asymmetric-key-specs.html#key-spec-ecc) does not support [[ref:Ed25519]]. At some point in the future, we can consider introducing support for key wrapping.
+:::
+
+Consumers of conformant [[ref:Web5 SDKs]] ****SHOULD**** be able to provide their own `KeyManager` impelementations, if desired.
+
+### JSON Schema
+
+[[ref:JSON-SCHEMA]] support is ****REQUIRED**** for multiple features throughout [[ref:Web5 SDKs]], including, but not limited to support with [[ref:verifiable credentials]], [VC JSON Schemas](#vc-json-schema), and for use in validating the data models this specification has defined.
```
Comment (KendallWeihe) [inline spec/spec.md:136]: should this be capitalized?

```suggestion
[[ref:JSON-SCHEMA]] support is ****REQUIRED**** for multiple features throughout [[ref:Web5 SDKs]], including, but not limited to support with [[ref:Verifiable Credentials]], [VC JSON Schemas](#vc-json-schema), and for use in validating the data models this specification has defined.
```

File: spec/spec.md (hunk 5 above)
Comment (KendallWeihe) [reply spec/spec.md:162]: okay so we should change this to 

```suggestion
| [[ref:did:web]] | ✅ | ✅ | - |
```

yeah?

perhaps we should add a column for `Publish`?

File: spec/spec.md [hunk 11]
```diff
@@ -0,0 +1,601 @@
...
+#### Supported DID Methods
+
+Web5 supports the following DID methods:
+
+| Method | Creation | Resolution | Note |
+|--------|----------|------------|------|
+| [[ref:did:web]] | ❌ | ✅ | - |
+| [[ref:did:jwk]] | ✅ | ✅ | - |
+| [[ref:did:dht]] | ✅ | ✅ | This is the "default" method for [[ref:Web5 SDKs]]. |
+| [[ref:did:key]] | ⚠️ | ⚠️ | This method has been implemented in both Kotlin and TypeScript, with no plans for support in other languages. |
```
Comment (KendallWeihe) [inline spec/spec.md:165]: I would probably just remove this line altogether

The new KT SDK (from Rust Core) won't have this

File: spec/spec.md (hunk 5 above)
Comment (decentralgabe) [reply spec/spec.md:162]: good call, yes

File: spec/spec.md (hunk 9 above)
Comment (decentralgabe) [reply spec/spec.md:104]: excellent, I'll add these

**Comment by decentralgabe** [comment]
@KendallWeihe @nitro-neal @frankhinek @mistermoe I have incorporated all feedback. I'd like to get this merged and publishing as-is. 

Then I will work on a follow up PR for aligning normative statements (MUSTs) to test vectors we have, in addition to removing the README and GH-pageifying the spec.

**Review by KendallWeihe — APPROVED** [review]

**Review by nitro-neal — APPROVED** [review]
Looks great! The big thing I would want to see in the future is a "Recommended `vc-jwt` verification and parsing .

Basically this part:
https://github.com/TBD54566975/web5-rs/blob/main/crates/web5/src/credentials/verifiable_credential_1_1.rs#L271

where a vc jwt is still valid if it has the jwt parts but not the internal vc parts, like it is valid if it hass `iss` and now vc.issuer (and when we build it internally we add vc.issuer from the `iss`

Basically this part
image

Maybe an example section of a "minimum viable vc-jwt" and a "full featured vc jwt"

**Comment by decentralgabe** [comment]
thanks @nitro-neal I added links to that section of the spec and an issue marker for #134

**Review by frankhinek — APPROVED** [review]

**Review by mistermoe — APPROVED** [review]
